[JSC] Unify 32-bit and 64-bit BBQJIT::addF64Copysign

mikhailramalho opened the pull request:
#### 442597c67d70bad57c53f18754e6630988c1759a
<pre>
[JSC] Unify 32-bit and 64-bit BBQJIT::addF64Copysign
<a href="https://bugs.webkit.org/show_bug.cgi?id=298957">https://bugs.webkit.org/show_bug.cgi?id=298957</a>

Reviewed by Yusuke Suzuki.

This PR unifies the 32-bit and 64-bit versions of BBQJIT::addF64Copysign, using
fp instructions.

In most cases, we were converting things to scalar, doing the copy sign, and
converting back to fp. By using fp instructions, the code for all platforms is
simplified.

The change to MacroAssemblerX86_64.h was required because absDouble on x86 it’s
implemented as ANDNOT with a -0.0 mask loaded into the destination register. If
src == dst, the ANDNOT zeros the register, so we use fpTempRegister to prevent
that. Also since we are not using shift/rotation on x86 anymore, we don&apos;t need
to clobber shiftRCX.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::absDouble):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpsllq_i8rr):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::F64CopysignHelper): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign): Deleted.

Canonical link: <a href="https://commits.webkit.org/301080@main">https://commits.webkit.org/301080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81ab9abc83365b07affcbc627b8a32a570c9fe72

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33335 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129573 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75032 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93452 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62023 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125875 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74070 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33549 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28191 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73059 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115083 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104275 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28402 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132292 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121461 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49873 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101950 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106243 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101851 "Found 3 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47177 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25372 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46647 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19571 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49729 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55489 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151718 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49196 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38798 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52548 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50878 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mikhailramalho commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -1278,12 +1278,17 @@ class MacroAssemblerX86_64 : public AbstractMacroAssembler<Assembler> {
 
     void absDouble(FPRegisterID src, FPRegisterID dst)
     {
```
These changes were required because of the
```
m_jit.absDouble(resultLocation.asFPR(), resultLocation.asFPR());
```
on the new `BBQJIT::addF64Copysign`

Constellation replied:
This is problematic since it is using fpTempRegister while AbsDouble is listed in Air opcode list.
You need to do much more work here,
1. Lower AbsDouble to actual sequence of AndDouble etc. in ARM64 / ARM and make AbsDouble as macro in Air
2. Use fpTempRegister here after that

Also, why not just using fpTempRegister for constant place? So you can just use three operand vandnpd_rrr below.

mikhailramalho replied:
We changed the implementation so it doesn't use the fpTempRegister anymore, WDYT?

mikhailramalho committed:
[JSC] Unify 32-bit and 64-bit BBQJIT::addF64Copysign
https://bugs.webkit.org/show_bug.cgi?id=298957

Reviewed by Yusuke Suzuki.

This PR unifies the 32-bit and 64-bit versions of BBQJIT::addF64Copysign, using
fp instructions.

In most cases, we were converting things to scalar, doing the copy sign, and
converting back to fp. By using fp instructions, the code for all platforms is
simplified.

The change to MacroAssemblerX86_64.h was required because absDouble on x86 it’s
implemented as ANDNOT with a -0.0 mask loaded into the destination register. If
src == dst, the ANDNOT zeros the register, so we use fpTempRegister to prevent
that. Also since we are not using shift/rotation on x86 anymore, we don't need
to clobber shiftRCX.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::absDouble):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpsllq_i8rr):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::F64CopysignHelper): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Copysign): Deleted.

Canonical link: https://commits.webkit.org/301080@main
